ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -254,20 +277,47 @@ applyHelper ::
   -> ShelleyBlock era
   -> Ticked (LedgerState (ShelleyBlock era))
   -> m (LedgerState (ShelleyBlock era))
-applyHelper f cfg blk (TickedShelleyLedgerState _ oldShelleyState) = do
-    newShelleyState <- f globals oldShelleyState (shelleyBlockRaw blk)
+applyHelper f cfg blk TickedShelleyLedgerState{
+                          untickedShelleyLedgerTransition
+                        , tickedShelleyLedgerState
+                        } = do
+    newShelleyState <- f globals tickedShelleyLedgerState (shelleyBlockRaw blk)
 
     return ShelleyLedgerState {
-        shelleyLedgerTip        = NotOrigin ShelleyTip {
-                                      shelleyTipBlockNo = blockNo   blk
-                                    , shelleyTipSlotNo  = blockSlot blk
-                                    , shelleyTipHash    = blockHash blk
-                                    }
-      , shelleyLedgerState      = newShelleyState
-      , shelleyLedgerTransition = ShelleyTransitionUnknown
+        shelleyLedgerTip = NotOrigin ShelleyTip {
+            shelleyTipBlockNo = blockNo   blk
+          , shelleyTipSlotNo  = blockSlot blk
+          , shelleyTipHash    = blockHash blk
+          }
+      , shelleyLedgerState =
+          newShelleyState
+      , shelleyLedgerTransition = ShelleyTransitionInfo {
+            shelleyAfterVoting =
+              if blockSlot blk < votingDeadline then
+                0
+              else
+                succ (shelleyAfterVoting untickedShelleyLedgerTransition)
+          }
       }
   where
     globals = shelleyLedgerGlobals cfg
+    swindow = SL.stabilityWindow globals
+
+    ei :: EpochInfo Identity
+    ei = SL.epochInfo (shelleyLedgerGlobals cfg)
+
+    -- The start of the next epoch is within the safe zone, always.
+    startOfNextEpoch :: SlotNo
+    startOfNextEpoch = runIdentity $ do
+        blockEpoch <- epochInfoEpoch ei (blockSlot blk)
+        let nextEpoch = succ blockEpoch
+        epochInfoFirst ei nextEpoch
+
+    -- The block must come in strictly before the voting deadline
+    -- See Fig 13, "Protocol Parameter Update Inference Rules", of the
+    -- Shelley specification.
+    votingDeadline :: SlotNo
```

Comment:
- The new `subSlots` is only used here in `votingDeadline`.
- `votingDeadline` is only used in a comparison.

Therefore we don't need `subSlots`. We can instead move the subtraction on the right of `<` to an addition on the left of it.

Food for thought, since that let's us avoid `subSlots` and the `assert` in its definition. (We'd technically convert that `assert` to an overflow check, but we haven't been doing those anywhere for Slot etc's Word64.)

Reply:
I take your point, but I think I prefer to keep it, to make sure that the code is more obviously conform spec. However, there _is_ a real bug here; it should be _twice_ the stability period :) Will fix that.